File: Harness Python/harness/runner.py
import asyncio
import time
from dataclasses import dataclass
from providers import REGISTRY
from evaluators import get_evaluator

@dataclass
class TestResult:
    test_id: str
    provider: str
    input: str
    output: str
    passed: bool
    score: float
    reason: str
    latency_ms: float
    error: str = ""

async def run_single(provider_name: str, test_case: dict) -> TestResult:
    provider = REGISTRY.get(provider_name)
    if not provider:
        return TestResult(
            test_id=test_case.get("id", "unknown"),
            provider=provider_name,
            input=test_case.get("input", ""),
            output="",
            passed=False,
            score=0.0,
            reason="Provider not found",
            latency_ms=0.0,
            error="Provider not found in registry"
        )
        
    start = time.monotonic()
    try:
        output = await provider.complete([{"role": "user", "content": test_case["input"]}])
        latency = (time.monotonic() - start) * 1000
        
        evaluator = get_evaluator(test_case["eval_type"])
        result = await evaluator.evaluate(output, test_case)
        
        return TestResult(
            test_id=test_case.get("id", "unknown"),
            provider=provider_name,
            input=test_case.get("input", ""),
            output=output,
            passed=result.passed,
            score=result.score,
            reason=result.reason,
            latency_ms=round(latency, 1),
        )
    except Exception as e:
        latency = (time.monotonic() - start) * 1000
        return TestResult(
            test_id=test_case.get("id", "unknown"),
            provider=provider_name,
            input=test_case.get("input", ""),
            output="",
            passed=False,
            score=0.0,
            reason="",
            latency_ms=round(latency, 1),
            error=str(e),
        )

async def run_suite(
    test_cases: list[dict],
    providers: list[str],
    concurrency: int = 5
) -> list[TestResult]:
    semaphore = asyncio.Semaphore(concurrency)

    async def bounded(p, tc):
        async with semaphore:
            return await run_single(p, tc)

    tasks = [bounded(p, tc) for p in providers for tc in test_cases]
    return await asyncio.gather(*tasks)

File: Harness Typescript/src/harness/runner.ts
import pLimit from 'p-limit';
import { BaseProvider } from '../providers/base.js';
import { EVALUATORS } from '../evaluators/index.js';

export interface TestCase {
  id: string;
  prompt?: string;
  input?: string;
  evaluator?: string;
  eval_type?: string;
  expected?: any;
  expected_contains?: any;
}

export interface TestResult {
  provider: string;
  test_id: string;
  passed: boolean;
  score: number;
  latency_ms: number;
  reason: string;
  error: string;
}

export async function runSuite(testCases: TestCase[], providers: BaseProvider[]): Promise<TestResult[]> {
  const limit = pLimit(5);
  const results: TestResult[] = [];

  const tasks = [];

  for (const testCase of testCases) {
    for (const provider of providers) {
      tasks.push(limit(async () => {
        const start = Date.now();
        const result: TestResult = {
          provider: provider.name,
          test_id: testCase.id,
          passed: false,
          score: 0.0,
          latency_ms: 0,
          reason: "",
          error: ""
        };

        try {
          const content = testCase.prompt || testCase.input || "";
          const output = await provider.complete([{ role: "user", content: content }]);
          result.latency_ms = Date.now() - start;

          const evaluatorName = testCase.evaluator || testCase.eval_type || "exact_match";
          const expected = testCase.expected !== undefined ? testCase.expected : testCase.expected_contains;
          const evaluator = EVALUATORS[evaluatorName];
          
          if (!evaluator) {
            result.error = `Evaluator '${evaluatorName}' not found`;
          } else {
            const evalResult = await evaluator.evaluate(output, expected);
            result.passed = evalResult.passed;
            result.score = evalResult.score;
            result.reason = evalResult.reason || "";
          }
        } catch (e) {
          result.latency_ms = Date.now() - start;
          result.error = (e as Error).message;
        }

        results.push(result);
      }));
    }
  }

  await Promise.all(tasks);
  return results;
}
